### ISSUE TITLE
Wrong commas in select after applying "sqlfluff fix"

### ISSUE THREAD

**Issue by jan-benisek** [sqlfluff/sqlfluff#2306]
## Expected Behaviour
I expect that this should not change much, maybe `tabs`/`spaces`, but this is otherwise a valid `SELECT`.
```sql
select
    schema::varchar as table_schema,
    table::varchar as table_name,
    usename::varchar as user_name,
    starttime::timestamp as query_time,
    querytxt::varchar as query_txt
from  foo
...
```

## Observed Behaviour
After `sqlfluff fix --force --dialect 'redshift' foo/bar.sql` the commas get messed up.

```sql
select
	schema::varchar as table_schema, ,
	usename::varchar as user_name
	starttime::timestamp as query_time,
	querytxt::varchar as query_txt,
	table::varchar as table_name
from foo
```

## Steps to Reproduce
`sqlfluff fix --force --dialect 'redshift' foo/bar.sql`

## Dialect
Redshift

## Version
sqlfluff, version 0.9.1
Python 3.8.12
dbt: 1.0.1
Plugins:
  - redshift: 1.0.0
  - postgres: 1.0.1

## Configuration
[sqlfluff:core]
templater = dbt
sql_file_exts = .sql
exclude_rules = L011, L031, L042

[sqlfluff:templater:dbt]
project_dir = foo/
profile = bar

[sqlfluff:indentation]
indented_joins = false
indented_using_on = true
template_blocks_indent = false
indent_unit = tab

[sqlfluff:templater]
unwrap_wrapped_queries = true

[sqlfluff:rules]
tab_space_size = 4
max_line_length = 80
ignore_comment_lines = false
indent_unit = tab
comma_style = trailing

[sqlfluff:rules:L010]
capitalisation_policy = lower

[sqlfluff:rules:L014]
extended_capitalisation_policy = lower
unquoted_identifiers_policy = all

[sqlfluff:rules:L028]
single_table_references = consistent

[sqlfluff:rules:L029]
unquoted_identifiers_policy = aliases
quoted_identifiers_policy = aliases

[sqlfluff:rules:L030]
capitalisation_policy = lower

[sqlfluff:rules:L040]
capitalisation_policy = lower

**Comment by jpy-git** [comment]
Hi @jan-benisek,

This is because you're SQL statement doesn't parse in SQLFluff (you can see if you run `sqlfluff parse --dialect redshift foo/bar.sql`). Specifically this is because `TABLE` is a reserved keyword, are you sure that `TABLE` (unquoted) is actually a valid column name?
If it is then we can just move it to unreserved keywords 😄 

(@barrywhart another example of why we shouldn't be attempting to apply any fixes on unparsable code)

**Comment by barrywhart** [comment]
@jpy-git: +1 for disabling that behavior. If nobody addresses it before then, let's discuss in the next maintainer's meeting (February 1st).

**Comment by tunetheweb** [comment]
I’m OK with disabling `fix`. Do not think we should disable `lint` though.

**Comment by jan-benisek** [comment]
> Hi @jan-benisek,
> 
> This is because you're SQL statement doesn't parse in SQLFluff (you can see if you run `sqlfluff parse --dialect redshift foo/bar.sql`). Specifically this is because `TABLE` is a reserved keyword, are you sure that `TABLE` (unquoted) is actually a valid column name? If it is then we can just move it to unreserved keywords 😄
> 
> (@barrywhart another example of why we shouldn't be attempting to apply any fixes on unparsable code)

Yes, it is unfortunately a column name. Is there any way I can fix that?

**Comment by jpy-git** [comment]
@jan-benisek yeah it's a quick fix if you want to make it. You just need to move `TABLE` in `src/sqlfluff/dialects/dialect_redshift_keywords.py` from the `redshift_reserved_keywords` to the `redshift_unreserved_keywords` (similar PR we did the other day for reference: #2310).

**Comment by tunetheweb** [comment]
This could cause a number of other issues though so might not be quite as simple. `TABLE` is kind of an important word in SQL! 😉

However looking [here](https://docs.aws.amazon.com/redshift/latest/dg/r_pg_keywords.html) it says `TABLE` **IS** a reserved word and [according to this](https://docs.aws.amazon.com/redshift/latest/dg/r_names.html) "standard identifiers... must... Not be a reserved SQL key word."

So I'd prefer NOT to make any changes in SQLFluff's RedShift dialect to cater for this to be honest.

Are you definitely using RedShift? Or some variant that's close to it? Perhaps it;'s a bug in RedShift that allows this and it shouldn't really? And it might get fixed in the future?

Can you change your SQL to this (I don't know if this is valid SQL in RedShift):

```sql
"table"::varchar as table_name
```

Or, this which will currently allow it to `lint` and `fix`, though as per above, this might change in the future:

```sql
    table::varchar as table_name, -- noqa: PRS
```

**Comment by jpy-git** [comment]
Yeah I had similar thoughts. No harm in seeing what happens if you move it though, worst case it fails and we close the PR 🤷

Also a lot of those redshift reserved keywords turned out to be unreserved in #2310 (confirmed by @rpr-ableton), so not sure how much I trust the docs 😅

**Comment by rpr-ableton** [comment]
I wouldn't trust the Redshift official documentation blindly; it's quite often incomplete (if not straight up misleading).

In the tests I ran in #2310 and #2318 , I did not include tests on column names, this might be a different set than for table/schema/function names..?

I can try to dedicate some time later this week, if there's interest, to do more in-depth testing in order to explore all possible corner cases.

**Comment by jan-benisek** [comment]
> However looking here it says TABLE IS a reserved word and according to this "standard identifiers... must... Not be a reserved SQL key word."


Oh well that's funny, because it is a column name of one of the admin tables
```sql
SELECT      i.DATABASE,
                   i.SCHEMA,
                   i.TABLE,
                   i."TABLE" AS tst --also works
FROM        svv_table_info i
```

@tunetheweb I like both of your options to fix it, but that entails going through all the sql queries, and we have a lot. I'd rather that SQLFluff does nothing. Or even better, after `lint` and `fix`, is there a process which checks if this is still a valid code?

Thanks for the feedback, I will see what makes the most sense.

**Comment by tunetheweb** [comment]
I just moved `TABLE` from reserved to unreserved and test suite still seems to pass. So that is indeed an option.

**Comment by rpr-ableton** [comment]
@tunetheweb FWIW both `create temp table table as (select 1);` and `with table as (select 1) select * from table;` fail in Redshift.

That indicates to me that there might be a need to introduce more granularity then just reserved/unreserved? Maybe something similar to what exists in the PG dialect?

**Comment by tunetheweb** [comment]
Postgres has [a number of types](https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/dialects/dialect_postgres_keywords.py) (`"reserved"`, `"non-reserved"`, `"not-keyword"`, `"reserved-(can-be-function-or-type)"`), but this is for future use (like required here!) and they basically are still reserved or not.

So basically at the moment SQLFluff can't handle semi-reserved words and a keyword has to be either reserved or not.

I've tested and moving `TABLE` to unreserved word list does seem to work, but then it will allow above examples as `valid`?

My preference is still to keep this as a reserved word - `TABLE` is so key to SQL - until we can handle semi-reserved words. Which is back to the workarounds suggested above, which I appreciate might mean a lot of edits to your codebase to support 😞

**Comment by jpy-git** [comment]
Actually coming back to this, I was surprised to see that `TABLE` is already an unreserved keyword in ANSI (and has been for the last 2 years). If we were to see issues from moving `TABLE` to unreserved then we would've almost certainly seen it in ANSI.
I personally have no issues moving the Postgres `TABLE` keyword to unreserved, as although it's not strictly unreserved in every situation, I think it's best to lean on the permissive side in the meantime if it's not causing issues.
